presto-hive/src/main/java/com/facebook/presto/hive/HiveStagingFileCommitter.java

```diff
@@ -0,0 +1,86 @@
+
+import com.facebook.presto.hive.HdfsEnvironment.HdfsContext;
+import com.facebook.presto.hive.PartitionUpdate.StagingFileInfo;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.SchemaTableName;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_FILESYSTEM_ERROR;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class HiveStagingFileCommitter
+        implements StagingFileCommitter
+{
+    private final HdfsEnvironment hdfsEnvironment;
+    private final Executor renameExecutor;
+
+    public HiveStagingFileCommitter(
+            HdfsEnvironment hdfsEnvironment,
+            Executor renameExecutor)
+    {
+        this.hdfsEnvironment = requireNonNull(hdfsEnvironment, "hdfsEnvironment is null");
+        this.renameExecutor = requireNonNull(renameExecutor, "renameExecutor is null");
+    }
+
+    @Override
+    public void commitFiles(ConnectorSession session, SchemaTableName tableName, List<PartitionUpdate> fragments)
+    {
+        List<Future<?>> fileRenameFutures = new ArrayList<>();
+        for (PartitionUpdate partitionUpdate : fragments) {
+            Path path = partitionUpdate.getWritePath();
+            HdfsContext context = new HdfsContext(session, tableName.getSchemaName(), tableName.getTableName());
+            FileSystem fileSystem;
+            try {
+                fileSystem = hdfsEnvironment.getFileSystem(context, path);
+            }
+            catch (IOException e) {
+                throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error moving data files to final location. Error listing directory %s", path), e);
+            }
+            for (StagingFileInfo stagingFileInfo : partitionUpdate.getStagingFileInfos()) {
+                fileRenameFutures.add(CompletableFuture.runAsync(() -> renameFile(fileSystem, path, stagingFileInfo), renameExecutor));
+            }
+        }
+        for (Future<?> fileRenameFuture : fileRenameFutures) {
+            getFutureValue(fileRenameFuture, PrestoException.class);
+        }
+    }
+
+    private void renameFile(FileSystem fileSystem, Path path, StagingFileInfo stagingFileInfo)
+    {
+        Path source = new Path(path, stagingFileInfo.getStagingFileName());
+        Path target = new Path(path, stagingFileInfo.getTargetFileName());
+        try {
+            if (fileSystem.exists(target) || !fileSystem.rename(source, target)) {
+                throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Error moving data files from %s to final location %s", source, target));
```

Comment:
nit: Maybe extract it to a method, to do not copy the error message

Reply:
It's hard to extract  `PrestoException` construction because when `IOException` is caught, we want to include it into `PrestoException` as well. I'll add a private method to construct the error message.